### PR TITLE
Verification of the wss.pem certificate in mod_sofia application code. Display correct reason why sofia profile failed to create

### PR DESCRIPTION
**Pull request by figaro2015**
https://github.com/signalwire/freeswitch/issues/1033

### CODE REVIEW

File: src/mod/endpoints/mod_sofia/sofia.c [hunk 1]
```diff
@@ -3185,6 +3192,37 @@ void *SWITCH_THREAD_FUNC sofia_profile_thread_run(switch_thread_t *thread, void
 		profile->tls_verify_in_subjects = su_strlst_dup_split((su_home_t *)profile->nua, profile->tls_verify_in_subjects_str, "|");
 	}
 
+	ssl_ctx = SSL_CTX_new((SSL_METHOD *)ssl_method);
```
Comment (crienzo) [inline src/mod/endpoints/mod_sofia/sofia.c:3195]: This must be destroyed when finished.  Any use of SSL must check if `HAVE_OPENSSL` is defined and the correct version of openssl is available.

**Review by crienzo — CHANGES_REQUESTED** [review]

File: src/mod/endpoints/mod_sofia/sofia.c (hunk 1 above)
Comment (mjerris) [reply src/mod/endpoints/mod_sofia/sofia.c:3195]: also can we just make this a core helper function that does all what chris said properly and returns a result and use that function here

**Review by mjerris — CHANGES_REQUESTED** [review]

**Comment by figaro2015** [comment]
Based on the comments provided I added a check for HAVE_OPENSSL pre-processor directive define. And the release of previously allocated SSL_CTX object.
I don't believe any additional OpenSSL version check is required as 'SSLv23_server_method', according to the OpenSSL documentation, is version flexible and support all currently available version of SSL/TLS protocols.

"SSLv23_method(), SSLv23_server_method(), SSLv23_client_method()
These are the general-purpose version-flexible SSL/TLS methods. The actual protocol version used will be negotiated to the highest version mutually supported by the client and the server. The supported protocols are SSLv2, SSLv3, TLSv1, TLSv1.1 and TLSv1.2. Most applications should use these method, and avoid the version specific methods described below"

Please take a look at certificate validation in MOD_VERTO. It's pretty similar to what I am doing in MOD_SOFIA.

As far as creating a separate core library helper utility - there are at least 4 or 5 places throughout FreeSWITCH modules where similar certificates validation/verification performs. Therefore, creating such a centralized utility function, to really serve the purpose, would involve modifications in all those places.

File: src/mod/endpoints/mod_sofia/sofia.c [hunk 2]
```diff
@@ -3267,9 +3313,16 @@ void *SWITCH_THREAD_FUNC sofia_profile_thread_run(switch_thread_t *thread, void
...
 	if (!profile->nua) {
-		switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_ERROR, "Error Creating SIP UA for profile: %s (%s)\n"
-						  "The likely causes for this are:\n" "1) Another application is already listening on the specified address.\n"
-						  "2) The IP the profile is attempting to bind to is not local to this system.\n", profile->name, profile->bindurl);
+		if (!ssl_error) {
+			switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_ERROR, "Error Creating SIP UA for profile: %s (%s)\n"
+							  "The likely causes for this are:\n" "1) Another application is already listening on the specified address.\n"
+							  "2) The IP the profile is attempting to bind to is not local to this system.\n", profile->name, profile->bindurl);
+		}
+		else {
```
Comment (crienzo) [inline src/mod/endpoints/mod_sofia/sofia.c:3321]: } else {

**Comment by crienzo** [comment]
Just a minor nit for me.  @mjerris is this OK otherwise?

File: src/mod/endpoints/mod_sofia/sofia.c (hunk 2 above)
Comment (figaro2015) [reply src/mod/endpoints/mod_sofia/sofia.c]: Corrected if/else curly braces formatting. Checked in.

**Review by crienzo — APPROVED** [review]

**Pull request merged**